### ISSUE TITLE
Segfault when using passive viewer with flexcomp in the loaded xml

### ISSUE THREAD

**Issue by lbusellato** [google-deepmind/mujoco#1219]
Hi,

I'm a Master's student in Robotics and I'm trying to use MuJoCo (v. 3.0.1) w/ Python for simulating the contact of an UR5e robot's end-effector with a soft, deformable object.

To do so, I've used the scene provided in the Menagerie for the UR5, with minimal tweaks:

<details>
  <summary>Scene XML</summary>

```XML
<mujoco model="ur5e scene">
  <compiler angle="radian" meshdir="meshes/ur5e" autolimits="true"/>

  <option integrator="implicitfast"/>

  <option timestep="0.002">
    <flag warmstart="enable"></flag>
  </option>

  <option impratio="10" />

  <extension>
    <plugin plugin="mujoco.elasticity.solid"/>
  </extension>

  <statistic center="0.3 0 0.4" extent="0.8"/>

  <visual>
    <headlight diffuse="0.6 0.6 0.6" ambient="0.1 0.1 0.1" specular="0 0 0"/>
    <rgba haze="0.15 0.25 0.35 1"/>
    <global azimuth="120" elevation="-20"/>
  </visual>

  <asset>
    <texture type="skybox" builtin="gradient" rgb1="0.3 0.5 0.7" rgb2="0 0 0" width="512" height="3072"/>
    <texture type="2d" name="groundplane" builtin="checker" mark="edge" rgb1="0.2 0.3 0.4" rgb2="0.1 0.2 0.3"
      markrgb="0.8 0.8 0.8" width="300" height="300"/>
    <material name="groundplane" texture="groundplane" texuniform="true" texrepeat="5 5" reflectance="0.2"/>
  </asset>

  <default>
    <default class="ur5e">
      <material specular="0.5" shininess="0.25"/>
      <joint axis="0 1 0" range="-6.28319 6.28319" armature="0.1"/>
      <!-- position actuators -->
      <!-- <general gaintype="fixed" biastype="affine" ctrlrange="-6.2831 6.2831" gainprm="2000"
      biasprm="0 -2000 -400"
        forcerange="-150 150"/> -->
      <default class="size3">
        <default class="size3_limited">
          <joint range="-3.1415 3.1415" />
          <!-- <general ctrlrange="-3.1415 3.1415"/> -->
        </default>
      </default>
      <default class="size1">
        <general gainprm="500" biasprm="0 -500 -100" forcerange="-28 28"/>
      </default>
      <default class="visual">
        <geom type="mesh" contype="0" conaffinity="0" group="2"/>
      </default>
      <default class="collision">
        <geom type="capsule" group="3"/>
        <default class="eef_collision">
          <geom type="cylinder"/>
        </default>
      </default>
      <site size="0.001" rgba="0.5 0.5 0.5 0.3" group="4"/>
    </default>
  </default>

  <asset>
    <material class="ur5e" name="black" rgba="0.033 0.033 0.033 1"/>
    <material class="ur5e" name="jointgray" rgba="0.278 0.278 0.278 1"/>
    <material class="ur5e" name="linkgray" rgba="0.82 0.82 0.82 1"/>
    <material class="ur5e" name="urblue" rgba="0.49 0.678 0.8 1"/>

    <mesh file="base_0.obj"/>
    <mesh file="base_1.obj"/>
    <mesh file="shoulder_0.obj"/>
    <mesh file="shoulder_1.obj"/>
    <mesh file="shoulder_2.obj"/>
    <mesh file="upperarm_0.obj"/>
    <mesh file="upperarm_1.obj"/>
    <mesh file="upperarm_2.obj"/>
    <mesh file="upperarm_3.obj"/>
    <mesh file="forearm_0.obj"/>
    <mesh file="forearm_1.obj"/>
    <mesh file="forearm_2.obj"/>
    <mesh file="forearm_3.obj"/>
    <mesh file="wrist1_0.obj"/>
    <mesh file="wrist1_1.obj"/>
    <mesh file="wrist1_2.obj"/>
    <mesh file="wrist2_0.obj"/>
    <mesh file="wrist2_1.obj"/>
    <mesh file="wrist2_2.obj"/>
    <mesh file="wrist3.obj"/>
    <mesh file="end_effector.stl"/>
  </asset>

  <worldbody>
    <light pos="0 0 1.5" dir="0 0 -1" directional="true"/>
    <geom name="floor" size="0 0 0.05" type="plane" material="groundplane"/>
    <body name="base" quat="1 0 0 1" childclass="ur5e">
      <inertial mass="4.0" pos="0 0 0" diaginertia="0.00443333156 0.00443333156 0.0072"/>
      <geom mesh="base_0" material="black" class="visual"/>
      <geom mesh="base_1" material="jointgray" class="visual"/>
      <body name="shoulder_link" pos="0 0 0.163">
        <inertial mass="3.7" pos="0 0 0" diaginertia="0.0102675 0.0102675 0.00666"/>
        <joint name="shoulder_pan_joint" class="size3" axis="0 0 1"/>
        <geom mesh="shoulder_0" material="urblue" class="visual"/>
        <geom mesh="shoulder_1" material="black" class="visual"/>
        <geom mesh="shoulder_2" material="jointgray" class="visual"/>
        <geom class="collision" size="0.06 0.06" pos="0 0 -0.04"/>
        <body name="upper_arm_link" pos="0 0.138 0" quat="1 0 1 0">
          <inertial mass="8.393" pos="0 0 0.2125" diaginertia="0.133886 0.133886 0.0151074"/>
          <joint name="shoulder_lift_joint" class="size3"/>
          <geom mesh="upperarm_0" material="linkgray" class="visual"/>
          <geom mesh="upperarm_1" material="black" class="visual"/>
          <geom mesh="upperarm_2" material="jointgray" class="visual"/>
          <geom mesh="upperarm_3" material="urblue" class="visual"/>
          <geom class="collision" pos="0 -0.04 0" quat="1 1 0 0" size="0.06 0.06"/>
          <geom class="collision" size="0.05 0.2" pos="0 0 0.2"/>
          <body name="forearm_link" pos="0 -0.131 0.425">
            <inertial mass="2.275" pos="0 0 0.196" diaginertia="0.0311796 0.0311796 0.004095"/>
            <joint name="elbow_joint" class="size3_limited"/>
            <geom mesh="forearm_0" material="urblue" class="visual"/>
            <geom mesh="forearm_1" material="linkgray" class="visual"/>
            <geom mesh="forearm_2" material="black" class="visual"/>
            <geom mesh="forearm_3" material="jointgray" class="visual"/>
            <geom class="collision" pos="0 0.08 0" quat="1 1 0 0" size="0.055 0.06"/>
            <geom class="collision" size="0.038 0.19" pos="0 0 0.2"/>
            <body name="wrist_1_link" pos="0 0 0.392" quat="1 0 1 0">
              <inertial mass="1.219" pos="0 0.127 0" diaginertia="0.0025599 0.0025599 0.0021942"/>
              <joint name="wrist_1_joint" class="size1"/>
              <geom mesh="wrist1_0" material="black" class="visual"/>
              <geom mesh="wrist1_1" material="urblue" class="visual"/>
              <geom mesh="wrist1_2" material="jointgray" class="visual"/>
              <geom class="collision" pos="0 0.05 0" quat="1 1 0 0" size="0.04 0.07"/>
              <body name="wrist_2_link" pos="0 0.127 0">
                <inertial mass="1.219" pos="0 0 0.1" diaginertia="0.0025599 0.0025599 0.0021942"/>
                <joint name="wrist_2_joint" axis="0 0 1" class="size1"/>
                <geom mesh="wrist2_0" material="black" class="visual"/>
                <geom mesh="wrist2_1" material="urblue" class="visual"/>
                <geom mesh="wrist2_2" material="jointgray" class="visual"/>
                <geom class="collision" size="0.04 0.06" pos="0 0 0.04"/>
                <geom class="collision" pos="0 0.02 0.1" quat="1 1 0 0" size="0.04 0.04"/>
                <body name="wrist_3_link" pos="0 0 0.1">
                  <inertial mass="0.1889" pos="0 0.0771683 0" quat="1 0 0 1"
                    diaginertia="0.000132134 9.90863e-05 9.90863e-05"/>
                  <joint name="wrist_3_joint" class="size1"/>
                  <geom material="linkgray" mesh="wrist3" class="visual"/>
                  <geom class="eef_collision" pos="0 0.08 0" quat="1 1 0 0" size="0.04 0.02"/>
                  <site name="attach_site" pos="0 0.1 0" quat="1 0 0 1"/>
                  <body name="end_effector" pos="0 0.1 0" quat="1 1 0 0">
                    <inertial mass="0.05" pos="0 0.0771683 0" quat="1 0 0 1"
                      diaginertia="0.000132134 9.90863e-05 9.90863e-05"/>
                    <geom material="jointgray" mesh="end_effector" class="visual"/>
                    <geom class="eef_collision" pos="0 0 -0.0512" quat="1 0 0 1" size="0.02 0.0512"/>
                    <site name="eef_site" pos="0 0 -0.1024" quat="1 0 0 1"/>
                    
                  </body>
                </body>
              </body>
            </body>
          </body>
        </body>
      </body>
    </body>
  </worldbody>

  <actuator>
    <motor class="size3" name="shoulder_pan" joint="shoulder_pan_joint"/>
    <motor class="size3" name="shoulder_lift" joint="shoulder_lift_joint"/>
    <motor class="size3_limited" name="elbow" joint="elbow_joint"/>
    <motor class="size1" name="wrist_1" joint="wrist_1_joint"/>
    <motor class="size1" name="wrist_2" joint="wrist_2_joint"/>
    <motor class="size1" name="wrist_3" joint="wrist_3_joint"/>
  </actuator>

  <sensor>
    <force name="eef_force" site="eef_site" cutoff="50" />
    <torque name="eef_torque" site="eef_site" cutoff="10" />
  </sensor>

  <include file="deformable.xml"/>
</mujoco>

```

</details>

In the last line, I include another XML which contains the definition of the flexcomp representing the deformable object:

<details>
  <summary>Deformable object XML</summary>

```XML
 <mujoco model="deformable">
   <worldbody>
   <flexcomp type="grid" count="5 5 5" spacing=".02 .02 .02" pos="0.5 0.5 0.04"
             radius=".0" rgba="0 .7 .7 1" name="softbody" dim="3" mass="1">
      <contact condim="3" solref="0.01 0.5" solimp=".95 .99 .0001" selfcollide="none"/>
      <edge damping="1"/>
      <plugin plugin="mujoco.elasticity.solid">
         <config key="poisson" value="0.1"/>
         <!--Units are in Pa (SI)-->
         <config key="young" value="1e5"/>
      </plugin>
   </flexcomp>
</worldbody>
</mujoco>

```

</details>

I'm then using the passive viewer to visualize the execution of some trajectory. I do this more or less like this:

```
import mujoco
import mujoco.viewer

m = mujoco.MjModel.from_xml_path('scene.xml')
d = mujoco.MjData(m)

with mujoco.viewer.launch_passive(m, d) as viewer:
    while viewer.is_running():
        mujoco.mj_step(m, d)
        viewer.sync()

```

As soon as I execute the script however, I get a segfault. I used faulthandler to try and locate the issue, and this was the output:

<details>
  <summary>Faulthandler output</summary>
Fatal Python error: Segmentation fault

Thread 0x00007f1489f2d700 (most recent call first):
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/mujoco/viewer.py", line 200 in _reload
  File "/usr/lib/python3.8/threading.py", line 870 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap

Current thread 0x00007f148972c700 (most recent call first):
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/mujoco/viewer.py", line 395 in _launch_internal
  File "/usr/lib/python3.8/threading.py", line 870 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f1486f2b700 (most recent call first):
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/Xlib/protocol/display.py", line 561 in send_and_recv
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/Xlib/protocol/rq.py", line 1380 in reply
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/Xlib/protocol/rq.py", line 1368 in __init__
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/Xlib/ext/record.py", line 219 in __init__
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/Xlib/ext/record.py", line 238 in enable_context
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/pynput/_util/xorg.py", line 407 in _run
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/pynput/keyboard/_xorg.py", line 545 in _run
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/pynput/_util/__init__.py", line 210 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f148d298740 (most recent call first):
  File "/usr/lib/python3.8/threading.py", line 302 in wait
  File "/usr/lib/python3.8/queue.py", line 170 in get
  File "/home/buse/mujoco-venv/lib/python3.8/site-packages/mujoco/viewer.py", line 473 in launch_passive
  File "/home/buse/Desktop/CoLeCT_sim/colect_sim/env/ur5_env.py", line 261 in __init__
  File "/home/buse/Desktop/CoLeCT_sim/main.py", line 10 in \<module>
Segmentation fault (core dumped)
</details>

This does not happen when removing the import of the flexcomp, in that case the viewer pops up and the simulation plays out as expected. I've tried looking at the documentation as well as other issues (#1193 had the same problem, but the commit that supposedly fixed the issue didn't), without luck. Can someone point me in the right direction? Thanks.

**Comment by kevinzakka** [comment]
HI @lbusellato, does the model load in the native simulate window?

**Comment by lbusellato** [comment]
Hello @kevinzakka, yes, I can use simulate to load, separately, either the ur5 or the deformable cube, but not with them together in the scene.

**Comment by kevinzakka** [comment]
I lowered the timestep to `0.001` and it's much better behaved. Can you try that on your end?

**Comment by lbusellato** [comment]
Tried that with simulate and it works. Oddly, simulate works fine also with the old timestep now, even though I did not change anything else (besides rebooting my system). 

In any case, viewer still segfaults. The window pops up for a split second, and I can see the left/right ui tabs, while the middle part is black.

**Comment by giorgionicola** [comment]
Hi, I am the author of (https://github.com/google-deepmind/mujoco/issues/1193), I have downloaded the latest version of Mujoco 3.0.1 and fixed the segmentation error on all models except those using the plugin mujoco.elasticity.solid

I am reopening the issue

**Comment by lbusellato** [comment]
Hello, thanks all for the quick response. Unfortunately I am still having issues. I tried installing mujoco from github head (with `pip install mujoco --pre -f https://py.mujoco.org/`, and it installs version 3.0.2.dev586380774 without issue. When I launch the viewer however, I get the following error:

```
ERROR: mj_stackAlloc: insufficient memory: max = 0, available = 0, requested = 24 (ne = 0, nf = 0, nefc = 0, ncon = 0)

Press Enter to exit ...
```

Am I missing something obvious here? Thanks all again for the help.

**Comment by yuvaltassa** [comment]
Did you read [this](https://mujoco.readthedocs.io/en/latest/modeling.html#csize)?
TL;DR you want `<size memory="..."/>`.

That said, the `max = 0` bit in your printout is surprising, so maybe this is something else...

**Comment by lbusellato** [comment]
Hello @yuvaltassa, tried with `<size memory="100"/>`, got the following:

```
ValueError: Error: engine error: mj_stackAlloc: insufficient memory: max = 100, available = 72, requested = 768 (ne = 0, nf = 0, nefc = 0, ncon = 0)
```

Tried increasing it further to `<size memory="10000"/>`, got the following:

```
ValueError: Error: engine error: mj_stackAlloc: insufficient memory: max = 10000, available = 760, requested = 9216 (ne = 0, nf = 0, nefc = 0, ncon = 0)
```

Since my system has 12GB of available RAM to work with, I wanted to increase it a lot more to see what would happen by setting `<size memory="4G"/>`, and I get this:

```
ERROR: mj_stackAlloc: insufficient memory: max = 0, available = 0, requested = 24 (ne = 0, nf = 0, nefc = 0, ncon = 0)

Press Enter to exit ...
```
Tried also the other suffixes (K and M), and none seems to set the max.

**Comment by yuvaltassa** [comment]
Extremely strange!

1. If you avoid the K/M/G suffixes and just add zeros, it works?
2. This is with the exact models you posted above? Or did you tweak them since then? If yes I'll try to reproduce.
3. Please give the details of your system: OS, local languages, processor, whatever might be releveant.
